# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/wasm/WasmInstance.cpp
# Commit: 89db2f360b8f
# Full Hash: 89db2f360b8fd68c398c70a8c556b92b2058ab00
# Author: Ryan Hunt <rhunt@eqrion.net>
# Date: 2022-03-24 21:54:20
# Regressor Bug: 1642412
# File Overlap Count: 7
# Description:
#   Bug 1642412 - wasm: Implement a 'precise' post-write barrier in baseline. r=jseward
#   
#   Add a post-write barrier which will remove a previous store buffer entry if the
#   new value stored in the field would not require it. This requires the previous
#   value to be loaded before the store and propagated to the post-write barrier
# ==============================================================================

diff -r eabb888a0851 -r 89db2f360b8f js/src/wasm/WasmInstance.cpp
--- a/js/src/wasm/WasmInstance.cpp	Thu Mar 24 17:56:52 2022 +0000
+++ b/js/src/wasm/WasmInstance.cpp	Thu Mar 24 17:56:52 2022 +0000
@@ -1158,6 +1158,15 @@
       reinterpret_cast<JSObject**>(location));
 }
 
+/* static */ void Instance::postBarrierPrecise(Instance* instance,
+                                               JSObject** location,
+                                               JSObject* prev) {
+  MOZ_ASSERT(SASigPostBarrierPrecise.failureMode == FailureMode::Infallible);
+  MOZ_ASSERT(location);
+  JSObject* next = *location;
+  JSObject::postWriteBarrier(location, prev, next);
+}
+
 /* static */ void Instance::postBarrierFiltering(Instance* instance,
                                                  gc::Cell** location) {
   MOZ_ASSERT(SASigPostBarrierFiltering.failureMode == FailureMode::Infallible);